src/lib/loaders/batchLoader.ts

```diff
@@ -69,26 +111,40 @@ export const batchLoader = ({
   if (!ENABLE_RESOLVER_BATCHING) {
     return singleLoader ? singleLoader : multipleLoader
   }
-  const dl = new DataLoader(keys => {
-    let groupedKeys = groupKeys(keys as any)
-
-    return Promise.all(
-      groupedKeys.map(keys => {
-        if (keys.id.length === 1 && singleLoader) {
-          return singleLoader(keys.id[0])
-        } else {
-          return multipleLoader(keys)
-        }
-      })
-    ).then(data => {
-      const normalizedResults = data.map((queriedGroup, groupIndex) => {
-        return groupedKeys[groupIndex].id.map(
-          id => queriedGroup.find(r => r._id === id) || defaultResult
+  const dl = new DataLoader(
+    (keys: Key[]) => {
+      const normalKeys = normalizeKeys(keys)
+      const [keyGroups, groupedKeys] = groupKeys(normalKeys)
+
+      return Promise.all(
+        groupedKeys.map(compactKeyForRequest).map(keys => {
+          if (keys.id.length === 1 && singleLoader) {
+            return singleLoader(keys.id[0])
+          } else {
+            return multipleLoader(keys)
+          }
+        })
+      ).then((data: Array<GravityResult | GravityResult[]>) => {
+        const normalizedData = data.map(
+          datum => (Array.isArray(datum) ? datum : [datum])
         )
+
+        return normalKeys.map(key => {
+          const group = getKeyGroup(key)
+          const groupIndex = keyGroups.indexOf(group)
```

Comment:
Perhaps I'm not evaluating all of this correctly in my mind, but I wonder if it's possible to add the key to the data coming out of the data loaders [in the code block starting on line 120] and that removing the need to find the right group here again. Nothing earth-shattering in terms of optimisations, but perhaps a bit simpler.

Reply:
The point of this section is to make sure that we return results in the exact order that they were passed into the data loader. The only data guarantee that I have is that the groups that I process will be returned in the same order. Results inside the group could be out of order. Adding a key to the results wouldn't really make sense because we already have a key (of sorts) in the results in the form of `_id` or `_slug`. 

I'm not sure. I know this is complicated. I'm open for ideas to make it simpler, I'm just not seeing how shifting this logic and mixing it in with the fetching would do that.

Reply:
I'm not really seeing anything concrete, so I'd leave it as-is and later on see if there's opportunity to iterate. Perhaps we can pair, when it all works as you want it to.